### PR TITLE
Swap and slot reduction

### LINKED ISSUE #5552: `numberOfErsatzValidators` not honoured if slot-reduction + swap happens in same epoch
# Problem encountered
If the network parameter that controls the number of ersatz validators is reduced in the same epoch that an ersatz validator is promoted, we can be left with a nodeset where `actual-number-or-ersatz > t.numberOfErsatzValidators`

I have added a unittest on this branch that flexes it: [swap-and-slot-reduction](https://github.com/vegaprotocol/vega/compare/swap-and-slot-reduction?expand=1)

and have seen it as an unstable test in this system-test run: https://jenkins.ops.vega.xyz/blue/organizations/jenkins/common%2Fsystem-tests/detail/system-tests/4225/pipeline/277

# Observed behaviour
Sometimes when the network parameter for the number of ersatz validators is reduced, the number of actual ersatz validators never changes.

# Expected behaviour
The value of the network parameter is always honoured.

# Definition of Done
>ℹ️ Not every issue will need every item checked, however, every item on this list should be properly considered and actioned to meet the [DoD](https://github.com/vegaprotocol/vega/blob/develop/DEFINITION_OF_DONE.md).

**Before Merging**
- [ ] Code refactored to meet SOLID and other code design principles
- [ ] Code is compilation error, warning, and hint free
- [ ] Carry out a basic happy path end-to-end check of the new code
- [ ] All APIs are documented so auto-generated documentation is created
- [ ] All bug recreation steps can be followed without presenting the original error/bug
- [ ] All Unit, Integration and BVT tests are passing
- [ ] Implementation is peer reviewed (coding standards, meeting acceptance criteria, code/design quality)
- [ ] Create [front end](https://github.com/vegaprotocol/token-frontend/issues) or [console](https://github.com/vegaprotocol/console/issues) tickets with feature labels (should be done when starting the work if dependencies known i.e. API changes)

**After Merging**
- [ ] Move development ticket to `Done` if there is **NO** requirement for new [system-tests](https://github.com/vegaprotocol/system-tests/issues)
- [ ] Resolve any issues with broken system-tests
- [ ] Create [documentation](https://github.com/vegaprotocol/documentation/issues) tickets with feature labels if functionality has changed, or is a new feature

### PR DESCRIPTION
**Pull request by wwestgarth**
closes #5552

### CODE REVIEW

**Review by jeremyletang — APPROVED** [review]

File: validators/validator_set.go [hunk 1]
```diff
@@ -298,6 +298,11 @@ func (t *Topology) applyPromotion(performanceScore, rankingScore map[string]num.
 		}
 	}
 
+	// demoted tendermint validators are also ersatz so we need to add them
+	for _, id := range demotedFromTM {
+		ersatzValidators = append(ersatzValidators, t.validators[id])
+	}
+
 	// demote from ersatz to pending due to more ersatz than slots allowed
 	ersatzValidators, waitingListValidators, _ = demoteDueToLackOfSlots(ersatzValidators, waitingListValidators, ValidatorStatusErsatz, ValidatorStatusPending, t.numberOfErsatzValidators, int64(t.currentBlockHeight+1))
```
Comment (wwestgarth) [inline validators/validator_set.go:307]: The one thing I am unsure of is whether we should be then removing the `demotedFromTM` validators from `ersatzValidators` before we call promote before to stop a TM -> pending. Its unclear to me from the spec whether in: 
"If a node that submitted the transaction to join and satisfies all other conditions and has a higher score than the lowest scoring ersatz validator (scaled up by the incumbent factor)"

`lowest scoring ersatz validator` means the lowest scoring before the tendermint <-> ersatz promotions are applied, or after the tendermint <-> ersatz promotions.

File: validators/validator_set.go [hunk 2]
```diff
@@ -298,6 +298,11 @@ func (t *Topology) applyPromotion(performanceScore, rankingScore map[string]num.
 		}
 	}
 
+	// demoted tendermint validators are also ersatz so we need to add them
+	for _, id := range demotedFromTM {
```
Comment (ze97286) [inline validators/validator_set.go:302]: maybe I'm misremembering this, but won't they already be there by construction? IIRC promote would already return them with the correct status in remainingValidators and then the iteration in line 293 would already put them in ersatz. what am I missing?

Comment (wwestgarth) [reply validators/validator_set.go:302]: No, in `promote` its only when the number of slots reduces that we move validators from seriesA into seriesB (and `removedFromSeriesA` is left empty). When a true "demotion" happens the demoted validator is returned in `removedFromSeriesA` but not returned in `seriesB`.

Which feels intentional?

Comment (ze97286) [reply validators/validator_set.go:302]: it feels messy mostly, but that's on me

**Review by ze97286 — APPROVED** [review]

**Pull request merged**